source/common/ssl/context_config_impl.cc

```diff
@@ -41,18 +42,43 @@ ContextConfigImpl::ContextConfigImpl(const envoy::api::v2::auth::CommonTlsContex
           Config::DataSource::read(config.validation_context().crl(), true)),
       certificate_revocation_list_path_(
           Config::DataSource::getPath(config.validation_context().crl())),
-      cert_chain_(
-          config.tls_certificates().empty()
-              ? ""
-              : Config::DataSource::read(config.tls_certificates()[0].certificate_chain(), true)),
+      cert_chain_([&config, &secret_manager] {
+        if (!config.tls_certificates().empty()) {
+          return Config::DataSource::read(config.tls_certificates()[0].certificate_chain(), true);
+        } else if (!config.tls_certificate_sds_secret_configs().empty()) {
+          auto static_secret =
+              secret_manager.staticSecret(config.tls_certificate_sds_secret_configs()[0].name());
+          if (static_secret == nullptr) {
+            throw EnvoyException(
+                fmt::format("Static secret is not defined: {}",
+                            config.tls_certificate_sds_secret_configs()[0].name()));
+          }
+          return static_secret->certificateChain();
+        } else {
+          return std::string("");
+        }
+      }()),
       cert_chain_path_(
           config.tls_certificates().empty()
               ? ""
               : Config::DataSource::getPath(config.tls_certificates()[0].certificate_chain())),
-      private_key_(
-          config.tls_certificates().empty()
-              ? ""
-              : Config::DataSource::read(config.tls_certificates()[0].private_key(), true)),
+      private_key_([&config, &secret_manager] {
+        if (!config.tls_certificates().empty()) {
```

Comment:
Can this logic be lifted out into a utility and used both here and above? it looks the same.

Reply:
Those are same structure but calling method are different. 
function might requires either an argument to choose certificate chain/private key or function pointer.
I am not sure that is efficient.